tests/test_cli.py

```diff
@@ -3,54 +3,250 @@
+    else:
+        print('Process Error Message:\n' + str(error))
+    return str(output)
+
+
+def cli_setup(runner):
+    result = runner.invoke(cli.setup, ['-c', VALID_DEVICECONNECTIONSTRING, '-g', 'iotedgetestingnow'])
+    assert 'Setup IoT Edge Simulator successfully' in result.output.strip()
+
+
+def cli_start_with_deployment(runner, deployment_file):
+    test_resources_dir = os.path.join('tests', 'test_compose_resources')
+    deployment_json_file_path = os.path.join(test_resources_dir, deployment_file)
+    result = runner.invoke(cli.start, ['-d', deployment_json_file_path])
+    assert 'IoT Edge Simulator has been started in solution mode' in result.output.strip()
+
+
+def get_device_id_from_device_connection_string():
+    device_id = re.findall(".*DeviceId=(.*);SharedAccessKey.*", VALID_DEVICECONNECTIONSTRING)[0]
+    return device_id
+
+
+def get_iothub_name_from_device_connection_string():
+    iothub_name = re.findall(".*HostName=(.*);DeviceId.*", VALID_DEVICECONNECTIONSTRING)[0].split('.')[0]
+    return iothub_name
+
+
+def invoke_module_method():
+    device_id = get_device_id_from_device_connection_string()
+    iothub_name = get_iothub_name_from_device_connection_string()
+    invoke_module_method_cmd = 'az iot hub invoke-module-method --device-id "' + device_id + \
+        '" --method-name "reset" --module-id "tempSensor" --hub-name "' + \
+        iothub_name + '" --login "' + VALID_IOTHUBCONNECTIONSTRING + '"'
+    output = start_process(invoke_module_method_cmd, True)
+    if '"status": 200' not in output:
+        raise "Failed to invoke module method."
+
+
+def monitor_d2c_message():
+    device_id = get_device_id_from_device_connection_string()
+    iothub_name = get_iothub_name_from_device_connection_string()
+    invoke_monitor_events_cmd = 'az iot hub monitor-events -n "' + iothub_name + \
+        '" -d "' + device_id + '" --login "' + VALID_IOTHUBCONNECTIONSTRING + '" -y -t 5'
+    output = start_process(invoke_monitor_events_cmd, True)
+    return output
+
+
+def cli_stop(runner):
+    result = runner.invoke(cli.stop)
+    assert not result.exception
+    assert result.exit_code == 0
+    assert 'IoT Edge Simulator has been stopped successfully' in result.output.strip()
+
+
+def reset_docker_environment():
+    start_process(['docker', 'network', 'rm', 'azure-iot-edge-dev'], False)
+    start_process(['docker', 'rmi', '-f', 'mcr.microsoft.com/azureiotedge-simulated-temperature-sensor:1.0'], False)
+    start_process(['docker', 'rmi', '-f', 'mcr.microsoft.com/azureiotedge-hub:1.0'], False)
+    start_process(['docker', 'rmi', '-f', 'hello-world'], False)
+
+
+def verify_docker_output(docker_cmd, expect_values):
+    result = start_process(docker_cmd, False)
+    print('Process result is: \n %s \n' % (result))
+    for expect_value in expect_values:
+        if expect_value in result:
+            print('%s is in process result.\n' % (expect_value))
+        else:
```

Comment:
We also need to make sure there are no errors in the output

Reply:
This method will check whether it throws error.

Reply:
If it throws error, output variable should be empty.

Reply:
Could you point me to the line number?

Reply:
Consider this scenario: the output contains our expected values, and also errors we haven't taken into consideration yet. WE should capture this errors.